File: bindings/wasm/tests/modes.test.js
"use strict";

// Operating-mode equivalence through the binding, over the golden corpus. The
// command protocol offers two ways to reach the same answer, and both must
// return the same bytes: `facts` is an alias of `build_context`, and `query`
// answers either against the context the handle stored from its last
// `build_context` or against a context passed inline. The core pins this in
// Rust (operating_modes.rs); this checks the boundary the WebAssembly
// core crosses. A missing corpus is a failure, not a skip.

const { test } = require("node:test");
const assert = require("node:assert");
const fs = require("node:fs");
const path = require("node:path");
const { Copilot } = require("../pkg-node/wickra_copilot_wasm.js");

const QUESTION = "what moved and why?";

function findGolden() {
  const g = path.resolve(__dirname, "..", "..", "..", "golden");
  return fs.existsSync(path.join(g, "specs")) ? g : null;
}

test("facts alias and inline query match the stored path", () => {
  const golden = findGolden();
  assert.ok(golden, "golden corpus not found");
  const feeds = JSON.parse(fs.readFileSync(path.join(golden, "feeds.json"), "utf8"));
  const specDir = path.join(golden, "specs");
  const files = fs.readdirSync(specDir).filter((f) => f.endsWith(".json")).sort();
  assert.ok(files.length > 0, "golden corpus not found");
  for (const file of files) {
    const spec = fs.readFileSync(path.join(specDir, file), "utf8");
    const expected = fs.readFileSync(path.join(golden, "expected", file), "utf8").trim();

    const stored = new Copilot(spec);
    const built = stored.command(JSON.stringify({ cmd: "build_context", feeds }));
    assert.strictEqual(built, expected, file);
    const fromStored = stored.command(JSON.stringify({ cmd: "query", question: QUESTION }));

    const fresh = new Copilot(spec);
    const fromInline = fresh.command(
      JSON.stringify({ cmd: "query", question: QUESTION, context: JSON.parse(built) }),
    );
    assert.strictEqual(fromInline, fromStored, file);

    const alias = new Copilot(spec);
    assert.strictEqual(alias.command(JSON.stringify({ cmd: "facts", feeds })), built, file);
  }
});

File: bindings/go/modes_test.go
package copilot

// Operating-mode equivalence through the binding, over the golden corpus. The
// command protocol offers two ways to reach the same answer, and both must
// return the same bytes: `facts` is an alias of `build_context`, and `query`
// answers either against the context the handle stored from its last
// `build_context` or against a context passed inline. The core pins this in
// Rust (operating_modes.rs); this checks the boundary the Go binding crosses.
// A missing corpus is a failure, not a skip.

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"
)

const modesQuestion = "what moved and why?"

func command(t *testing.T, c *Copilot, cmd map[string]any) string {
	t.Helper()
	b, err := json.Marshal(cmd)
	if err != nil {
		t.Fatal(err)
	}
	out, err := c.Command(string(b))
	if err != nil {
		t.Fatal(err)
	}
	return strings.TrimSpace(out)
}

func TestFactsAliasAndInlineQueryMatchTheStoredPath(t *testing.T) {
	g := goldenDir()
	if g == "" {
		t.Fatal("golden corpus not found")
	}
	feeds, err := os.ReadFile(filepath.Join(g, "feeds.json"))
	if err != nil {
		t.Fatal(err)
	}
	specs, err := filepath.Glob(filepath.Join(g, "specs", "*.json"))
	if err != nil || len(specs) == 0 {
		t.Fatal("golden corpus not found")
	}
	sort.Strings(specs)
	for _, specPath := range specs {
		name := filepath.Base(specPath)
		spec, err := os.ReadFile(specPath)
		if err != nil {
			t.Fatal(err)
		}
		expected, err := os.ReadFile(filepath.Join(g, "expected", name))
		if err != nil {
			t.Fatal(err)
		}

		stored, err := New(string(spec))
		if err != nil {
			t.Fatalf("%s: %v", name, err)
		}
		built := command(t, stored, map[string]any{"cmd": "build_context", "feeds": json.RawMessage(feeds)})
		if built != strings.TrimSpace(string(expected)) {
			t.Fatalf("%s: build_context does not match the blessed context", name)
		}
		fromStored := command(t, stored, map[string]any{"cmd": "query", "question": modesQuestion})
		stored.Close()

		fresh, err := New(string(spec))
		if err != nil {
			t.Fatalf("%s: %v", name, err)
		}
		fromInline := command(t, fresh, map[string]any{"cmd": "query", "question": modesQuestion, "context": json.RawMessage(built)})
		fresh.Close()
		if fromInline != fromStored {
			t.Fatalf("%s: an inline context answers differently from the stored one", name)
		}

		alias, err := New(string(spec))
		if err != nil {
			t.Fatalf("%s: %v", name, err)
		}
		viaFacts := command(t, alias, map[string]any{"cmd": "facts", "feeds": json.RawMessage(feeds)})
		alias.Close()
		if viaFacts != built {
			t.Fatalf("%s: facts differs from build_context", name)
		}
	}
}
